test/Test/Common.purs

```diff
@@ -0,0 +1,547 @@
+import Data.List.Lazy.NonEmpty as LNEL
+
+{-
+This is temporarily being used during development.
+It allows testing while still patching the API.
+This is passed as an additional argument to testCommon
+to indicate which container type is being tested, and
+lets us skip gaps that are currently implemented by `unsafeCrashWith`:
+
+Once fully supported by all containers, can replace with original assert.
+-}
+data SkipBroken
+  = SkipBrokenStrictCanEmpty
+  | SkipBrokenStrictNonEmpty
+  | SkipBrokenLazyCanEmpty
+  | SkipBrokenLazyNonEmpty
+  | RunAll
+
+derive instance eqSkipBroken :: Eq SkipBroken
+
+assertSkipHelper :: SkipBroken -> Array SkipBroken -> (_ -> Boolean) -> Effect Unit
+assertSkipHelper skip arr f =
+  case Array.elem skip arr of
+    true -> log "...skipped"
+    false -> assert $ f unit
+
+printContainerType :: String -> Effect Unit
+printContainerType str = do
+  log "--------------------------------"
+  log str
+  log "--------------------------------"
+
+printTestType :: String -> Effect Unit
+printTestType str = do
+  log $ "---- " <> str <> " Tests ----"
+
+class (
+  Alt c
+  , Applicative c
+  , Apply c
+  , Bind c
+  , Eq (c Int)
+  --, Eq1 c -- missing from NonEmptyList, LazyNonEmptyList
+  , Extend c
+  , Foldable c
+  , FoldableWithIndex Int c
+  , Functor c
+  , FunctorWithIndex Int c
+  , Monad c
+  , Ord (c Int)
+  --, Ord1 c -- missing from NonEmptyList, LazyNonEmptyList
+  , Semigroup (c Int)
+  , Show (c Int)
+  , Traversable c
+  , TraversableWithIndex Int c
+  , Unfoldable1 c
+) <= Common c where
+  concat :: forall a. c (c a) -> c a
+  concatMap :: forall a. forall b. (a -> c b) -> c a -> c b
+  -- Should basic list have a cons function wrapping the Cons constructor?
+  cons :: forall a. a -> c a -> c a
+  elemIndex :: forall a. Eq a => a -> c a -> Maybe Int
+  elemLastIndex :: forall a. Eq a => a -> c a -> Maybe Int
+  findIndex :: forall a. (a -> Boolean) -> c a -> Maybe Int
+  findLastIndex :: forall a. (a -> Boolean) -> c a -> Maybe Int
+  foldM :: forall m a b. Monad m => (b -> a -> m b) -> b -> c a -> m b
+  index :: forall a. c a -> Int -> Maybe a
+  intersect :: forall a. Eq a => c a -> c a -> c a
+  intersectBy :: forall a. (a -> a -> Boolean) -> c a -> c a -> c a
+  length :: forall a. c a -> Int
+  nubEq :: forall a. Eq a => c a -> c a
+  nubByEq :: forall a. (a -> a -> Boolean) -> c a -> c a
+  reverse :: c ~> c
+  singleton :: forall a. a -> c a
+  snoc :: forall a. c a -> a -> c a
+  toUnfoldable :: forall f a. Unfoldable f => c a -> f a
+  union :: forall a. Eq a => c a -> c a -> c a
+  unionBy :: forall a. (a -> a -> Boolean) -> c a -> c a -> c a
+  -- Types don't have to be all a
+  -- Todo - double check this requirement
+  unzip :: forall a b. c (Tuple a b) -> Tuple (c a) (c b)
+  zip :: forall a b. c a -> c b -> c (Tuple a b)
+  zipWith :: forall a b d. (a -> b -> d) -> c a -> c b -> c d
+  zipWithA :: forall a b d m. Applicative m => (a -> b -> m d) -> c a -> c b -> m (c d)
```

Comment:
Here's the example of APIs that are available to all list types. There should probably be more added here.

Reply:
Are we certain that all "list" types can (or should) implement the same APIs? It seems that at least some operations would be useful for some list types and not others -- for example, I'd imagine some operations don't make sense for a lazy list.

Reply:
I don't know the best place to have this discussion -- perhaps in the related "Patch API differences" conversation -- but my gut sense is that abstracting over list types in this way may have undesirable consequences for a) performance and b) error messages, and that we need to weigh that against the advantages of a unified interface.

Reply:
> Are we certain that all "list" types can (or should) implement the same APIs? It seems that at least some operations would be useful for some list types and not others -- for example, I'd imagine some operations don't make sense for a lazy list.

You are correct. I just linked to the most "universally common" snippet, but there are testing clusters in this PR that target just `Lazy` vs `"Strict"` lists or `"Emptiable"` vs `"NonEmpty"` lists. I think looking at the `UpdatedTests.purs` file in this PR is a good place to see how everything fits together.

> my gut sense is that abstracting over list types in this way may have undesirable consequences for a) performance and b) error messages, and that we need to weigh that against the advantages of a unified interface.

This abstraction is just for managing and running tests, so shouldn't affect end users of the library.
But regarding error messages during testing, I'd love to be able to use [`test-unit`](https://pursuit.purescript.org/packages/purescript-test-unit) (or similar) instead of `assert` so we can more easily see what's causing the failure.

Reply:
> I'd love to be able to use test-unit (or similar) instead of assert so we can more easily see what's causing the failure.

Core libraries only depend on other core libraries, so improvements to the testing situation would have to be added to `assert`; we won't be able to depend on other testing libraries, unfortunately.

Reply:
> I'd still like to hear your thoughts about what we should do when a library needs to diverge from the standard API -- how would you envision us handling that?

Currently, this is just a common API for this lists library, but to answer two related questions:

What if one of the sub-list types (e.g. `LazyNonEmpty`) needs to diverge from the common List API?
- In that case, the offending function is no longer common across all list types, and can be broken-out into whatever cluster that function is still common among.

What if we use a common API among multiple containers (e.g. Arrays and Lists) and some List functions need to diverge.
- Similar answer to the above. If those functions aren't common across all containers, then they can't be part of the common-container API, but they can be demoted to part of the common-list API.

Reply:
> I'd love to be able to use [`test-unit`](https://pursuit.purescript.org/packages/purescript-test-unit) (or similar) instead of `assert` so we can more easily see what's causing the failure.

I forgot we had `assertEqual`, which works fine for displaying diffs.